Node 17 built with openssl < 3 does not work with Vaadin 21.0.8+

### Description of the bug

I downloaded a starter Flow 22 project from https://start.vaadin.com/app/, but a clean project fails on `./mvnw` with:

```
This application has no explicit mapping for /error, so you are seeing this as a fallback.

Wed Dec 29 07:27:00 EST 2021
There was an unexpected error (type=Internal Server Error, status=500).
java.lang.IllegalStateException: Webpack exited prematurely
com.vaadin.flow.server.ServiceException: java.lang.IllegalStateException: Webpack exited prematurely
	at com.vaadin.flow.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1616)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1579)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:299)
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:109)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:711)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:459)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:353)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:313)
	at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:141)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:655)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:540)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:382)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:895)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1732)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Webpack exited prematurely
	at com.vaadin.base.devserver.AbstractDevServerRunner.doStartDevModeServer(AbstractDevServerRunner.java:220)
	at com.vaadin.base.devserver.AbstractDevServerRunner.runOnFutureComplete(AbstractDevServerRunner.java:172)
	at com.vaadin.base.devserver.AbstractDevServerRunner.lambda$new$0(AbstractDevServerRunner.java:163)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

There's no further/more detailed stack trace that I can find.

### Expected behavior

The application should start.

### Minimal reproducible example

1. Visit https://start.vaadin.com/app/
2. Under Settings, change Vaadin version to 22
3. Download project
4. Start with `./mvnw`
5. Browser launches with stacktrace

### Versions

- Vaadin / Flow version: 22.0.2
- Java version: 11
- OS version: macOS 12.1

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendTools.java
@@ -583,11 +583,6 @@ public class FrontendTools {
         }
         try {
             FrontendVersion foundNodeVersion = getNodeVersion();
-            if (foundNodeVersion.getMajorVersion() == 17
-                    && foundNodeVersion.getMinorVersion() == 0) {
-                throw new IllegalStateException(
-                        "Node version 17.0 is incompatible with webpack 4. Please upgrade to Node 17.1 or newer, or downgrade to Node 16.");
-            }
             FrontendUtils.validateToolVersion("node", foundNodeVersion,
                     SUPPORTED_NODE_VERSION, SHOULD_WORK_NODE_VERSION);
         } catch (UnknownVersionException e) {

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendToolsTest.java
@@ -375,18 +375,6 @@ public class FrontendToolsTest {
         Assert.assertTrue(file.exists());
     }
 
-    @Test
-    public void validateNodeAndNpmVersion_brokenNode17() throws Exception {
-        settings.setIgnoreVersionChecks(false);
-        tools = Mockito.spy(new FrontendTools(settings));
-
-        Mockito.when(tools.getNodeVersion())
-                .thenReturn(new FrontendVersion(17, 0));
-        Assert.assertThrows(IllegalStateException.class, () -> {
-            tools.validateNodeAndNpmVersion();
-        });
-    }
-
     @Test(expected = IllegalStateException.class)
     public void ensureNodeExecutableInHome_vaadinHomeNodeIsAFolder_throws()
             throws IOException {

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/AbstractDevServerRunner.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/AbstractDevServerRunner.java
@@ -299,15 +299,16 @@ public abstract class AbstractDevServerRunner implements DevModeHandler {
 
     /**
      * Defines the environment variables to use when starting the dev server.
-     * 
+     *
+     * @param frontendTools
+     *            frontend tools metadata
      * @param environment
      *            the environment variables to use
      */
-    protected void updateServerStartupEnvironment(FrontendVersion nodeVersion,
+    protected void updateServerStartupEnvironment(FrontendTools frontendTools,
             Map<String, String> environment) {
         environment.put("watchDogPort",
                 Integer.toString(getWatchDog().getWatchDogPort()));
-
     }
 
     /**
@@ -345,15 +346,7 @@ public abstract class AbstractDevServerRunner implements DevModeHandler {
         processBuilder.command(command);
 
         Map<String, String> environment = processBuilder.environment();
-        FrontendVersion nodeVersion;
-        try {
-            nodeVersion = tools.getNodeVersion();
-        } catch (UnknownVersionException e) {
-            getLogger().error("Unable to determine node version", e);
-            // Need to assume something..
-            nodeVersion = new FrontendVersion(16, 0, 0);
-        }
-        updateServerStartupEnvironment(nodeVersion, environment);
+        updateServerStartupEnvironment(tools, environment);
 
         try {
             Process process = processBuilder.redirectErrorStream(true).start();

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/WebpackHandler.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/WebpackHandler.java
@@ -16,6 +16,7 @@
 package com.vaadin.base.devserver;
 
 import java.io.File;
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -26,8 +27,8 @@ import java.util.regex.Pattern;
 import com.vaadin.base.devserver.DevServerOutputTracker.Result;
 import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.server.InitParameters;
+import com.vaadin.flow.server.frontend.FrontendTools;
 import com.vaadin.flow.server.frontend.FrontendUtils;
-import com.vaadin.flow.server.frontend.FrontendVersion;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -94,6 +95,9 @@ public final class WebpackHandler extends AbstractDevServerRunner {
     protected List<String> getServerStartupCommand(String nodeExec) {
         List<String> command = new ArrayList<>();
         command.add(nodeExec);
+        if (requiresOpenSslLegacyProvider(nodeExec)) {
+            command.add("--openssl-legacy-provider");
+        }
         command.add(getServerBinary().getAbsolutePath());
         command.add("--config");
         command.add(getServerConfig().getAbsolutePath());
@@ -119,10 +123,10 @@ public final class WebpackHandler extends AbstractDevServerRunner {
     }
 
     @Override
-    protected void updateServerStartupEnvironment(FrontendVersion nodeVersion,
+    protected void updateServerStartupEnvironment(FrontendTools frontendTools,
             Map<String, String> environment) {
-        super.updateServerStartupEnvironment(nodeVersion, environment);
-        if (!nodeVersion.isOlderThan(new FrontendVersion(17, 0, 0))) {
+        super.updateServerStartupEnvironment(frontendTools, environment);
+        if (requiresOpenSslLegacyProvider(frontendTools.getNodeExecutable())) {
             environment.put("NODE_OPTIONS", "--openssl-legacy-provider");
         }
     }
@@ -156,4 +160,27 @@ public final class WebpackHandler extends AbstractDevServerRunner {
                 DEFAULT_ERROR_PATTERN));
     }
 
+    private boolean requiresOpenSslLegacyProvider(String nodeExec) {
+        // Determine whether webpack requires Node.js to be started with the
+        // --openssl-legacy-provider parameter. This is a webpack 4 workaround
+        // of the issue https://github.com/webpack/webpack/issues/14532
+        // See: https://github.com/vaadin/flow/issues/12649
+
+        List<String> command = new ArrayList<>();
+        command.add(nodeExec);
+        ProcessBuilder processBuilder = new ProcessBuilder()
+                .directory(getProjectRoot())
+                .command(nodeExec, "-p", "crypto.createHash('md4')");
+        try {
+            Process process = processBuilder.start();
+            int errorLevel = process.waitFor();
+            return errorLevel != 0;
+        } catch (IOException | InterruptedException e) {
+            getLogger().error(
+                    "Exception while determining --openssl-legacy-provider "
+                            + "parameter requirement",
+                    e);
+        }
+        return false;
+    }
 }

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/WebpackHandler.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/WebpackHandler.java
@@ -165,9 +165,6 @@ public final class WebpackHandler extends AbstractDevServerRunner {
         // --openssl-legacy-provider parameter. This is a webpack 4 workaround
         // of the issue https://github.com/webpack/webpack/issues/14532
         // See: https://github.com/vaadin/flow/issues/12649
-
-        List<String> command = new ArrayList<>();
-        command.add(nodeExec);
         ProcessBuilder processBuilder = new ProcessBuilder()
                 .directory(getProjectRoot())
                 .command(nodeExec, "-p", "crypto.createHash('md4')");

--- a/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/WebpackHandler.java
+++ b/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/WebpackHandler.java
@@ -172,11 +172,18 @@ public final class WebpackHandler extends AbstractDevServerRunner {
             Process process = processBuilder.start();
             int errorLevel = process.waitFor();
             return errorLevel != 0;
-        } catch (IOException | InterruptedException e) {
+        } catch (IOException e) {
             getLogger().error(
-                    "Exception while determining --openssl-legacy-provider "
+                    "IO error while determining --openssl-legacy-provider "
                             + "parameter requirement",
                     e);
+        } catch (InterruptedException e) {
+            getLogger().error(
+                    "Interrupted while determining --openssl-legacy-provider "
+                            + "parameter requirement",
+                    e);
+            // re-interrupt the thread
+            Thread.currentThread().interrupt();
         }
         return false;
     }

